### ISSUE TITLE
Add CI (GitHub actions, Travis, Azure, etc)

### ISSUE THREAD

**Issue by kinow** [NCAS-CMS/cf-python#6]
Hi,

I executed the tests locally, after

1. cloning the repository
2. creating a virtual environment
3. running `pip install -e .`
4. `cd cf/test` and then `python run_tests.py`

But 5 tests failed. Not sure if I was missing some set up step?

But having Travis CI, or Azure Pipelines, or GitHub actions, or another CI testing my PR or branches in my own fork would have been helpful too to confirm my changes are not introducing new bugs.

Cheers
Bruno

**Comment by davidhassell** [comment]
Could you let me know which 5 tests failed? Thanks. I'll look into a CI for pull requests.

**Comment by kinow** [comment]
> Could you let me know which 5 tests failed? Thanks. I'll look into a CI for pull requests.

Sure. Just re-synced the repository, and executed the tests again with my virtual environment. Results after ~8 minutes with:

```bash
$ time python run_tests.py 
--------------------
CF-PYTHON TEST SUITE
--------------------
Run date: 2019-10-26 10:07:06.212272
Platform: Linux-4.15.0-66-generic-x86_64-with-debian-buster-sid
HDF5 library: 1.10.4
netcdf library: 4.6.3
udunits2 library: libudunits2.so.0
python: 3.7.3 /home/kinow/Development/python/workspace/cf-python/venv/bin/python
netCDF4: 1.5.3 /home/kinow/Development/python/workspace/cf-python/venv/lib/python3.7/site-packages/netCDF4/__init__.py
cftime: 1.0.4.2 /home/kinow/Development/python/workspace/cf-python/venv/lib/python3.7/site-packages/cftime/__init__.py
numpy: 1.17.3 /home/kinow/Development/python/workspace/cf-python/venv/lib/python3.7/site-packages/numpy/__init__.py
psutil: 5.6.3 /home/kinow/Development/python/workspace/cf-python/venv/lib/python3.7/site-packages/psutil/__init__.py
scipy: not available
matplotlib: not available
ESMF: not available
cfdm: 1.7.8 /home/kinow/Development/python/workspace/cf-python/venv/lib/python3.7/site-packages/cfdm/__init__.py
cfunits: 3.2.2 /home/kinow/Development/python/workspace/cf-python/venv/lib/python3.7/site-packages/cfunits/__init__.py
cfplot: not available
cf: 3.0.2 /home/kinow/Development/python/workspace/cf-python/cf/__init__.py

Running tests from /home/kinow/Development/python/workspace/cf-python/cf/test
test_create_field (setup_create_field.create_fieldTest) ... ok

----------------------------------------------------------------------
Ran 1 test in 0.270s

OK
test_AuxiliaryCoordinate__repr__str__dump (test_AuxiliaryCoordinate.AuxiliaryCoordinateTest) ... ok
test_AuxiliaryCoordinate_bounds (test_AuxiliaryCoordinate.AuxiliaryCoordinateTest) ... ok
test_AuxiliaryCoordinate_ceil (test_AuxiliaryCoordinate.AuxiliaryCoordinateTest) ... ok
test_AuxiliaryCoordinate_chunk (test_AuxiliaryCoordinate.AuxiliaryCoordinateTest) ... ok
test_AuxiliaryCoordinate_clip (test_AuxiliaryCoordinate.AuxiliaryCoordinateTest) ... ok
test_AuxiliaryCoordinate_close (test_AuxiliaryCoordinate.AuxiliaryCoordinateTest) ... ok
test_AuxiliaryCoordinate_count (test_AuxiliaryCoordinate.AuxiliaryCoordinateTest) ... ok
test_AuxiliaryCoordinate_cyclic (test_AuxiliaryCoordinate.AuxiliaryCoordinateTest) ... ok
test_AuxiliaryCoordinate_floor (test_AuxiliaryCoordinate.AuxiliaryCoordinateTest) ... ok
test_AuxiliaryCoordinate_insert_dimension (test_AuxiliaryCoordinate.AuxiliaryCoordinateTest) ... ok
test_AuxiliaryCoordinate_log_exp (test_AuxiliaryCoordinate.AuxiliaryCoordinateTest) ... ok
test_AuxiliaryCoordinate_mask_invalid (test_AuxiliaryCoordinate.AuxiliaryCoordinateTest) ... ok
test_AuxiliaryCoordinate_properties (test_AuxiliaryCoordinate.AuxiliaryCoordinateTest) ... ok
test_AuxiliaryCoordinate_rint (test_AuxiliaryCoordinate.AuxiliaryCoordinateTest) ... ok
test_AuxiliaryCoordinate_roll (test_AuxiliaryCoordinate.AuxiliaryCoordinateTest) ... ok
test_AuxiliaryCoordinate_round (test_AuxiliaryCoordinate.AuxiliaryCoordinateTest) ... ok
test_AuxiliaryCoordinate_sin_cos_tan (test_AuxiliaryCoordinate.AuxiliaryCoordinateTest) ... ok
test_AuxiliaryCoordinate_squeeze (test_AuxiliaryCoordinate.AuxiliaryCoordinateTest) ... ok
test_AuxiliaryCoordinate_transpose (test_AuxiliaryCoordinate.AuxiliaryCoordinateTest) ... ok
test_AuxiliaryCoordinate_trunc (test_AuxiliaryCoordinate.AuxiliaryCoordinateTest) ... ok
test_CellMeasure__repr__str__dump (test_CellMeasure.CellMeasureTest) ... ok
test_CellMeasure_identity (test_CellMeasure.CellMeasureTest) ... ok
test_CellMeasure_measure (test_CellMeasure.CellMeasureTest) ... ok
test_CellMethod__repr__str__ (test_CellMethod.CellMethodTest) ... ok
test_CellMethod_equals (test_CellMethod.CellMethodTest) ... ok
test_CellMethod_equivalent (test_CellMethod.CellMethodTest) ... ok
test_CellMethod_get_set_delete (test_CellMethod.CellMethodTest) ... ok
test_CoordinateReference__repr__str__dump (test_CoordinateReference.CoordinateReferenceTest) ... ok
test_CoordinateReference_canonical_units (test_CoordinateReference.CoordinateReferenceTest) ... ok
test_CoordinateReference_default_value (test_CoordinateReference.CoordinateReferenceTest) ... ok
test_CoordinateReference_equals (test_CoordinateReference.CoordinateReferenceTest) ... ok
test_CoordinateReference_get__getitem__ (test_CoordinateReference.CoordinateReferenceTest) ... ok
test_CoordinateReference_match (test_CoordinateReference.CoordinateReferenceTest) ... ok
test_Data_AUXILIARY_MASK (test_Data.DataTest) ... ok
test_Data_BINARY_AND_UNARY_OPERATORS (test_Data.DataTest) ... ok
test_Data_BROADCASTING (test_Data.DataTest) ... ok
test_Data_CachedArray (test_Data.DataTest) ... ok
test_Data_ERROR (test_Data.DataTest) ... ok
test_Data___contains__ (test_Data.DataTest) ... ok
test_Data___getitem__ (test_Data.DataTest) ... ok
test_Data___setitem__ (test_Data.DataTest) ... ok
test_Data__asdatetime__asreftime__isdatetime (test_Data.DataTest) ... ok
test_Data__collapse_SHAPE (test_Data.DataTest) ... ok
test_Data__float__ (test_Data.DataTest) ... ok
test_Data__int__ (test_Data.DataTest) ... ok
test_Data__len__ (test_Data.DataTest) ... ok
test_Data__round__ (test_Data.DataTest) ... ok
test_Data_all (test_Data.DataTest) ... ok
test_Data_any (test_Data.DataTest) ... ok
test_Data_argmax (test_Data.DataTest) ... ok
test_Data_array (test_Data.DataTest) ... ok
test_Data_asdata (test_Data.DataTest) ... ok
test_Data_binary_mask (test_Data.DataTest) ... ok
test_Data_ceil (test_Data.DataTest) ... ok
test_Data_clip (test_Data.DataTest) ... ok
test_Data_count (test_Data.DataTest) ... ok
test_Data_cumsum (test_Data.DataTest) ... ok
test_Data_datetime_array (test_Data.DataTest) ... ok
test_Data_datum (test_Data.DataTest) ... ok
test_Data_digitize (test_Data.DataTest) ... ok
test_Data_dumpd_loadd_dumps (test_Data.DataTest) ... ok
test_Data_exp (test_Data.DataTest) ... ok
test_Data_flatten (test_Data.DataTest) ... ok
test_Data_flip (test_Data.DataTest) ... ok
test_Data_floor (test_Data.DataTest) ... ok
test_Data_max (test_Data.DataTest) ... ok
test_Data_max_min_sum (test_Data.DataTest) ... ok
test_Data_mean (test_Data.DataTest) ... ok
test_Data_min (test_Data.DataTest) ... ok
test_Data_months_years (test_Data.DataTest) ... ok
test_Data_ndindex (test_Data.DataTest) ... ok
test_Data_outerproduct (test_Data.DataTest) ... ok
test_Data_rint (test_Data.DataTest) ... ok
test_Data_roll (test_Data.DataTest) ... ok
test_Data_round (test_Data.DataTest) ... ok
test_Data_sample_size (test_Data.DataTest) ... ok
test_Data_sd_var (test_Data.DataTest) ... ok
test_Data_section (test_Data.DataTest) ... ok
test_Data_sin_cos_tan (test_Data.DataTest) ... ok
test_Data_squeeze_insert_dimension (test_Data.DataTest) ... ok
test_Data_sum_of_weights_sum_of_weights2 (test_Data.DataTest) ... ok
test_Data_swapaxes (test_Data.DataTest) ... ok
test_Data_transpose (test_Data.DataTest) ... ok
test_Data_trunc (test_Data.DataTest) ... ok
test_Data_unique (test_Data.DataTest) ... ok
test_Data_varray (test_Data.DataTest) ... ok
test_Data_year_month_day_hour_minute_second (test_Data.DataTest) ... ok
test_Datetime (test_Datetime.DatetimeTest) ... ok
test_Datetime_Data (test_Datetime.DatetimeTest) ... ok
test_Datetime_dt2rt (test_Datetime.DatetimeTest) ... ok
test_Datetime_rt2dt (test_Datetime.DatetimeTest) ... ok
test_DimensionCoordinate__repr__str__dump (test_DimensionCoordinate.DimensionCoordinateTest) ... ok
test_DimensionCoordinate_binary_operation (test_DimensionCoordinate.DimensionCoordinateTest) ... ok
test_DimensionCoordinate_bounds (test_DimensionCoordinate.DimensionCoordinateTest) ... ok
test_DimensionCoordinate_cellsize (test_DimensionCoordinate.DimensionCoordinateTest) ... ok
test_DimensionCoordinate_convert_reference_time (test_DimensionCoordinate.DimensionCoordinateTest) ... ok
test_DimensionCoordinate_insert_dimension (test_DimensionCoordinate.DimensionCoordinateTest) ... ok
test_DimensionCoordinate_properties (test_DimensionCoordinate.DimensionCoordinateTest) ... ok
test_DimensionCoordinate_roll (test_DimensionCoordinate.DimensionCoordinateTest) ... ok
test_DomainAncillary (test_DomainAncillary.DomainAncillaryTest) ... ok
test_DomainAxis (test_DomainAxis.DomainAxisTest) ... ok
test_DomainAxis__repr__str (test_DomainAxis.DomainAxisTest) ... ok
test_Field_ATOL_RTOL (test_Field.FieldTest) ... ok
test_Field_AUXILIARY_MASK (test_Field.FieldTest) ... ok
test_Field_DATA (test_Field.FieldTest) ... ok
test_Field__add__ (test_Field.FieldTest) ... ok
test_Field__getitem__ (test_Field.FieldTest) ... ok
test_Field__gt__ (test_Field.FieldTest) ... ok
test_Field__mul__ (test_Field.FieldTest) ... ok
test_Field__setitem__ (test_Field.FieldTest) ... ok
test_Field_all (test_Field.FieldTest) ... ok
test_Field_allclose (test_Field.FieldTest) ... ok
test_Field_anchor (test_Field.FieldTest) ... ok
test_Field_any (test_Field.FieldTest) ... ok
test_Field_autocyclic (test_Field.FieldTest) ... ok
test_Field_auxiliary_coordinate (test_Field.FieldTest) ... ok
test_Field_axis (test_Field.FieldTest) ... ok
test_Field_bin (test_Field.FieldTest) ... ok
test_Field_cell_area (test_Field.FieldTest) ... ok
test_Field_cell_measure (test_Field.FieldTest) ... ok
test_Field_cell_method (test_Field.FieldTest) ... ok
test_Field_concatenate (test_Field.FieldTest) ... ok
test_Field_construct_key (test_Field.FieldTest) ... ok
test_Field_convert (test_Field.FieldTest) ... ok
test_Field_coordinate (test_Field.FieldTest) ... ok
test_Field_coordinate_reference (test_Field.FieldTest) ... ok
test_Field_cumsum (test_Field.FieldTest) ... ok
test_Field_dimension_coordinate (test_Field.FieldTest) ... ok
test_Field_direction (test_Field.FieldTest) ... ok
test_Field_domain_ancillary (test_Field.FieldTest) ... ok
test_Field_domain_axis (test_Field.FieldTest) ... ok
test_Field_domain_axis_position (test_Field.FieldTest) ... ok
test_Field_domain_mask (test_Field.FieldTest) ... ok
test_Field_equals (test_Field.FieldTest) ... ok
test_Field_field_ancillary (test_Field.FieldTest) ... ok
test_Field_flatten (test_Field.FieldTest) ... ok
test_Field_flip (test_Field.FieldTest) ... ok
test_Field_get_data_axes (test_Field.FieldTest) ... ok
test_Field_indices (test_Field.FieldTest) ... ok
test_Field_insert_dimension (test_Field.FieldTest) ... ok
test_Field_item (test_Field.FieldTest) ... ok
test_Field_mask_invalid (test_Field.FieldTest) ... ok
test_Field_match (test_Field.FieldTest) ... ok
test_Field_period (test_Field.FieldTest) ... ok
test_Field_replace_construct (test_Field.FieldTest) ... ok
test_Field_section (test_Field.FieldTest) ... ok
test_Field_squeeze (test_Field.FieldTest) ... ok
test_Field_transpose (test_Field.FieldTest) ... ok
test_Field_unsqueeze (test_Field.FieldTest) ... ok
test_Field_weights (test_Field.FieldTest) ... ok
test_Field_where (test_Field.FieldTest) ... ok
test_FieldAncillary (test_FieldAncillary.FieldAncillaryTest) ... ok
test_FieldList (test_FieldList.FieldTest) ... ok
test_FieldList__add__iadd__ (test_FieldList.FieldTest) ... ok
test_FieldList__contains__ (test_FieldList.FieldTest) ... ok
test_FieldList__getslice__ (test_FieldList.FieldTest) ... ok
test_FieldList__len__ (test_FieldList.FieldTest) ... ok
test_FieldList__mul__imul__ (test_FieldList.FieldTest) ... ok
test_FieldList__repr__ (test_FieldList.FieldTest) ... ok
test_FieldList_append_extend (test_FieldList.FieldTest) ... ok
test_FieldList_close (test_FieldList.FieldTest) ... ok
test_FieldList_concatenate (test_FieldList.FieldTest) ... ok
test_FieldList_copy (test_FieldList.FieldTest) ... ok
test_FieldList_count (test_FieldList.FieldTest) ... ok
test_FieldList_equals (test_FieldList.FieldTest) ... ok
test_FieldList_index (test_FieldList.FieldTest) ... ok
test_FieldList_insert_pop_remove (test_FieldList.FieldTest) ... ok
test_FieldList_reverse (test_FieldList.FieldTest) ... ok
test_FieldList_select (test_FieldList.FieldTest) ... ok
test_Maths (unittest.loader._FailedTest) ... ERROR
test_Partition (test_Partition.PartitionTest) ... ok
test_Query (test_Query.QueryTest) ... ok
test_Query_contains (test_Query.QueryTest) ... ok
test_Query_datetime1 (test_Query.QueryTest) ... ok
test_Query_dteq_dtne_dtge_dtgt_dtle_dtlt (test_Query.QueryTest) ... ok
test_Query_evaluate (test_Query.QueryTest) ... ok
test_Query_year_month_day_hour_minute_second (test_Query.QueryTest) ... ok
test_Field_regridc (test_Regrid.RegridTest) ... skipped 'Requires esmf package.'
test_Field_regrids (test_Regrid.RegridTest) ... skipped 'Requires esmf package.'
test_TimeDuration (test_TimeDuration.TimeDurationTest) ... ok
test_TimeDuration_arithmetic (test_TimeDuration.TimeDurationTest) ... ok
test_TimeDuration_bounds (test_TimeDuration.TimeDurationTest) ... ok
test_TimeDuration_interval (test_TimeDuration.TimeDurationTest) ... ok
test_TimeDuration_iso (test_TimeDuration.TimeDurationTest) ... ok
test_Timeduration__days_in_month (test_TimeDuration.TimeDurationTest) ... ok
test_aggregate (test_aggregate.aggregateTest) ... /home/kinow/Development/python/workspace/cf-python/cf/__init__.py:976: FutureWarning: elementwise comparison failed; returning scalar instead, but in the future will perform elementwise comparison
ok
test_COLLAPSE (test_collapse.FieldTest) ... ok
test_COLLAPSE_CLIMATOLOGICAL_TIME (test_collapse.FieldTest) ... ok
test_COLLAPSE_groups (test_collapse.FieldTest) ... ok
test_COLLAPSE_weights (test_collapse.FieldTest) ... ok
test_DSG_contiguous (test_dsg.DSGTest) ... ok
test_DSG_create_contiguous (test_dsg.DSGTest) ... ok
test_DSG_indexed (test_dsg.DSGTest) ... ok
test_DSG_indexed_contiguous (test_dsg.DSGTest) ... ok
test_EXTERNAL_READ (test_external.ExternalVariableTest) ... ok
test_EXTERNAL_WRITE (test_external.ExternalVariableTest) ... ok
test_GATHERING (test_gathering.DSGTest) ... ok
test_GATHERING_create (test_gathering.DSGTest) ... ok
test_GENERAL (test_general.generalTest) ... ok
test_PP_WGDOS_UNPACKING (test_pp.ppTest) ... ERROR
test_PP_load_stash2standard_name (test_pp.ppTest) ... ERROR
test_read_CDL (test_read_write.read_writeTest) ... /bin/sh: 1: ncdump: not found
ERROR
test_read_aggregate (test_read_write.read_writeTest) ... ok
test_read_directory (test_read_write.read_writeTest) ... ok
test_read_extra (test_read_write.read_writeTest) ... ok
test_read_pp (test_read_write.read_writeTest) ... ERROR
test_read_select (test_read_write.read_writeTest) ... ok
test_read_squeeze (test_read_write.read_writeTest) ... ok
test_read_write_format (test_read_write.read_writeTest) ... ok
test_read_write_netCDF4_compress_shuffle (test_read_write.read_writeTest) ... ok
test_read_write_unlimited (test_read_write.read_writeTest) ... ok
test_write_datatype (test_read_write.read_writeTest) ... ok
test_write_reference_datetime (test_read_write.read_writeTest) ... ok

======================================================================
ERROR: test_Maths (unittest.loader._FailedTest)
----------------------------------------------------------------------
ImportError: Failed to import test module: test_Maths
Traceback (most recent call last):
  File "/home/kinow/Development/python/anaconda3/lib/python3.7/unittest/loader.py", line 436, in _find_test_path
    module = self._get_module_from_name(name)
  File "/home/kinow/Development/python/anaconda3/lib/python3.7/unittest/loader.py", line 377, in _get_module_from_name
    __import__(name)
  File "/home/kinow/Development/python/workspace/cf-python/cf/test/test_Maths.py", line 8, in <module>
    from scipy.ndimage import convolve1d
ModuleNotFoundError: No module named 'scipy'


======================================================================
ERROR: test_PP_WGDOS_UNPACKING (test_pp.ppTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/kinow/Development/python/workspace/cf-python/cf/read_write/read.py", line 661, in _read_a_file
    ftype = file_type(filename)
  File "/home/kinow/Development/python/workspace/cf-python/cf/read_write/read.py", line 775, in file_type
    raise IOError("Can't determine format of file {}".format(filename))
OSError: Can't determine format of file /home/kinow/Development/python/workspace/cf-python/cf/test/wgdos_packed.pp

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/kinow/Development/python/workspace/cf-python/cf/test/test_pp.py", line 61, in test_PP_WGDOS_UNPACKING
    f = cf.read(self.ppfilename)[0]
  File "/home/kinow/Development/python/workspace/cf-python/cf/read_write/read.py", line 512, in read
    chunk=chunk)
  File "/home/kinow/Development/python/workspace/cf-python/cf/read_write/read.py", line 664, in _read_a_file
    raise Exception(error)
Exception: Can't determine format of file /home/kinow/Development/python/workspace/cf-python/cf/test/wgdos_packed.pp

======================================================================
ERROR: test_PP_load_stash2standard_name (test_pp.ppTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/kinow/Development/python/workspace/cf-python/cf/read_write/read.py", line 661, in _read_a_file
    ftype = file_type(filename)
  File "/home/kinow/Development/python/workspace/cf-python/cf/read_write/read.py", line 775, in file_type
    raise IOError("Can't determine format of file {}".format(filename))
OSError: Can't determine format of file /home/kinow/Development/python/workspace/cf-python/cf/test/wgdos_packed.pp

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/kinow/Development/python/workspace/cf-python/cf/test/test_pp.py", line 44, in test_PP_load_stash2standard_name
    f = cf.read(self.ppfilename)[0]
  File "/home/kinow/Development/python/workspace/cf-python/cf/read_write/read.py", line 512, in read
    chunk=chunk)
  File "/home/kinow/Development/python/workspace/cf-python/cf/read_write/read.py", line 664, in _read_a_file
    raise Exception(error)
Exception: Can't determine format of file /home/kinow/Development/python/workspace/cf-python/cf/test/wgdos_packed.pp

======================================================================
ERROR: test_read_CDL (test_read_write.read_writeTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/kinow/Development/python/workspace/cf-python/cf/test/test_read_write.py", line 304, in test_read_CDL
    shell=True, check=True)
  File "/home/kinow/Development/python/anaconda3/lib/python3.7/subprocess.py", line 487, in run
    output=stdout, stderr=stderr)
subprocess.CalledProcessError: Command 'ncdump /home/kinow/Development/python/workspace/cf-python/cf/test/test_file.nc > /tmp/tmp_ugj241m_cf-python_test' returned non-zero exit status 127.

======================================================================
ERROR: test_read_pp (test_read_write.read_writeTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/kinow/Development/python/workspace/cf-python/cf/read_write/read.py", line 661, in _read_a_file
    ftype = file_type(filename)
  File "/home/kinow/Development/python/workspace/cf-python/cf/read_write/read.py", line 775, in file_type
    raise IOError("Can't determine format of file {}".format(filename))
OSError: Can't determine format of file wgdos_packed.pp

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/kinow/Development/python/workspace/cf-python/cf/test/test_read_write.py", line 288, in test_read_pp
    p = cf.read('wgdos_packed.pp')[0]
  File "/home/kinow/Development/python/workspace/cf-python/cf/read_write/read.py", line 512, in read
    chunk=chunk)
  File "/home/kinow/Development/python/workspace/cf-python/cf/read_write/read.py", line 664, in _read_a_file
    raise Exception(error)
Exception: Can't determine format of file wgdos_packed.pp

----------------------------------------------------------------------
Ran 213 tests in 489.074s

FAILED (errors=5, skipped=2)

real	8m9.795s
user	8m2.401s
sys	0m7.713s
```

**Comment by sadielbartholomew** [comment]
Closing as a GH Actions workflow to run the full test suite has just been merged in #54. The workflow will in future be extended to include other metrics such as code coverage results, but the basic principle is now established & as per your motivation @kinow, will report test results based on the state of the PR. Thanks again.

**Comment by sadielbartholomew** [comment]
Closing as a GH Actions workflow to run the full test suite has just been merged in #54. The workflow will in future be extended to include other metrics such as code coverage results, but the basic principle is now established & as per your motivation @kinow, will report test results based on the state of the PR. Thanks again.

**Comment by kinow** [comment]
Thanks Sadie!!!
